### ISSUE TITLE
Parameter of type `never` causes return type inference to fail

### ISSUE THREAD

**Issue by AnyhowStep** [microsoft/TypeScript#33457]
**TypeScript Version:**  3.5.1

**Search Terms:**

parameter, never, return type inference

**Code**

```ts
type NeverF = (arg : never) => void;

//"y"
type Extends = NeverF extends (...args : any) => any ?
    "y" :
    "n"
;

type TsReturnType<T extends (...args : any) => any> =
    T extends (...args : any) => infer R ?
    R :
    "wtf"
;

//Expected: void
//Actual  : "wtf"
type r = TsReturnType<NeverF>;

//Expected: void
//Actual  : any
type r2 = ReturnType<NeverF>;
```

**Expected behavior:**

Both should resolve to `void`

**Actual behavior:**

Both resolve to the `false` branch

**Playground Link:** [Playground](http://www.typescriptlang.org/play/#code/C4TwDgpgBAchBuEBOAxKBeKAKAhkg5lAFxQB2CyAlBgHxTwD2AlgCYDcAUBwPTcBEIPh1CQoAUQAewCKRYBnDLAqooEKTPnYAdDrz4FJHKRDV0dIyCgB+DlDtQBfYrft9SQzsPDQAKnIBKEMAArkikPt4APD6q6rIKWDpaegZQFqbmxnToLnYxatLx2roEqem0UEykAGbIUP7WufXO9g4A7sDVHly8kpAAxtIsJIysPNwAgoPBOAA2diR8HV1eokiKfoEhYRGQkXCIqDSevRIDQyPMLONTIXMLacar0EgATIpboeFRB8gox0A)

**Related Issues:** 

Did a search for "parameter never in:title" and found nothing

-----

I'm using a function with a parameter of type `never` because I **only** care about the return type. And I do not want the user to invoke the function explicitly.

Given the following,
```ts
const query2 = myQuery.map((row) => {
  return {
    ...row,
    x : Math.random(),
  };
});
```

The type of `query2` should now be something like,
```ts
{
  /*snip*/
  mapDelegate : (row : never, connection : /*snip*/) => (
    { /*snip typeof row*/, x : number }
  ),
  /*snip*/
}
```

Then,
```ts
const query3 = query2.map((row) => {
  return {
    ...row,
    x : row.x + 3.141,
    y : "someNewValue",
  };
});
```

The type of `query3` should now be something like,
```ts
{
  /*snip*/
  mapDelegate : (row : never, connection : /*snip*/) => (
    { /*snip typeof row*/, x : number, y : string }
  ),
  /*snip*/
}
```

-----

I'm also using a `never` parameter because I want to save on `Emit time`. It sounds funny but literally every bit of text I shave off helps with compile times.

For large queries, the type of the `row` parameter could be huuuuuuggeeeee. Like, 100+ lines long.

So,

1. Only care about the return type
2. Not meant to be invoked externally
3. Save on emit time

**Comment by AnyhowStep** [comment]
My workaround at the moment,
```
type BetterReturnType<T extends (...args : never) => any> =
    T extends (...args : never) => infer R ?
    R :
    never
;
```

-----

```ts
type NeverF = (arg : never, arg1 : any, arg2 : never) => void;

//"y"
type Extends = NeverF extends (...args : any) => any ?
    "y" :
    "n"
;
type BetterReturnType<T extends (...args : never) => any> =
    T extends (...args : never) => infer R ?
    R :
    never
;

//Expected: void
//Actual  : void
type r = BetterReturnType<NeverF>;

//Expected: void
//Actual  : any
type r2 = ReturnType<NeverF>;
```
[Playground](http://www.typescriptlang.org/play/#code/C4TwDgpgBAchBuEBOAxKBeKAKAhkg5lAFxQB2CyANFHvgIzE2kjW0BMj5iSAlBgHxR4AewCWAEwDcAKGkB6OQCIQi6aEhQAogA9gEUuIDOGWBVRQIu-UewA6e7WMkczPukEuQUAPzSo-qGVFYj8AxVJVGXVoACEIYD0kACV4gFckUgAVcAgAHkyLKwNjLHtbR04zNw9mQXRQ-wLLPWK7BwInMiqBKFFSADNkKCSfBuGQgK7uaRl5OR1IAGM9cRIRCTmAQWXUnAAbfzWxcTUcqCQTOITkFOB0rJzcuG4UflmFBYhliFWhY62dvtDkwQKcNEgOJhbvdspAnmZXpIgA)

**Comment by jack-williams** [comment]
I think this one is subtle, and I'm not sure I have this right, but:

The type `(...args : any) => any` that is used in the extends type and constraint is classified as an _any signature_, and all signatures are unconditionally related to an any signature without any structural checking.

The problem is that it's not true that all signatures would be related to an any signature if you did the check structurally, specifically in the case where you use `never` because `never` is the only type that `any` is not related to.

The type `NeverF` is *not* structurally related to `(...args : any) => any` because the inputs are not related (contravariantly), however it gets a free pass because `(...args : any) => any` is classed as an _any signature_.

Things appear to work OK for `Extends` and when checking the constraint of `TsReturnType` because they both use any signatures, but when you infer and instantiate the extends type in `TsReturnType` you no longer have an any signature and therefore structural checking is applied --- this fails.

Here is a small explanation:
```ts
type NeverF = (arg : never) => void;
type AnySignature = (...args: any) => any;
type NotAnySignature = (...args: any) => unknown;

declare const neverF: NeverF;
let anySignature: AnySignature = neverF; // ok
let notAnySignature: NotAnySignature = neverF; // error, arguments are not related.
```

Changing the return type causes a type error to appear in the parameters because changing the return type no longer identifiers the signature as an _any signature._

**Comment by RyanCavanaugh** [comment]
@jack-williams is correct. `(...args: any[]) => any` is very special and won't behave to a type the appears to be structurally the same for certain purposes.

**Comment by AnyhowStep** [comment]
You're saying `ReturnType<>` shouldn't be changed to the implementation of `BetterReturnType<>` even though `BetterReturnType<>` appears to work better?

**Comment by AnyhowStep** [comment]
Also, just so I'm clear about the "intended" behavior of `extends (...args : any) => any`,

`(arg : never) => T` is assignable to `(...args : any) => any` but not really.

It behaves differently depending on whether it's used to constrain a type parameter, or used in a conditional type

**Comment by jack-williams** [comment]
> It behaves differently depending on whether it's used to constrain a type parameter, or used in a conditional type, right?

No it behaves the same. The difference is that in your conditional type the inference of `R` means you get a type that is altogether different: `(...args : any) => void`.

All signatures are related to the any signature, independently of where that question is asked in the checker.

It seems to me that if `(...arg : never) => infer R` can be used in `ReturnType` without causing any breaking changes that would be an improvement.

**Comment by AnyhowStep** [comment]
Was there a reason to make the false branch return `any` instead of `never`?

**Comment by AnyhowStep** [comment]
I tried it.

![image](https://user-images.githubusercontent.com/5655961/65169872-801c0880-da15-11e9-9046-ec92c49e6b16.png)

![image](https://user-images.githubusercontent.com/5655961/65169885-87431680-da15-11e9-9860-1bb12743b956.png)

I tried it on the playground,
```ts
/**
 * Notice the `extends (...args: any` part in the constraint.
 * And the `extends (...args: never` part in the conditional type.
 * 
 * This is intentional.
 */
type BetterReturnType<T extends (...args: any) => any> =
    T extends (...args: never) => infer R ? R : any;
type T19<T extends any[]> = BetterReturnType<(x: string, ...args: T) => T[]>;

//type x = number[][]
type x = T19<number[]>
```
[Playground](http://www.typescriptlang.org/play/#code/PQKhCgAIUg5B7ALgSwMYFNKIBaYAboAei6AdgCYDOkAFAHQMCGATgOaUBckjpAnnpAAOLRJGSksuSKnilKiZo3GI6UGAEEKk-ERIVq9Jm06RS6AG7pmA4c1Hjt02eWQpZjADZZeg9KuiQapAAKtjI1OFipHpupJ7+IMDgiD6YAELoiCTMAEqZAK7MpMGpADzBkLpkVLQMdCzsXDy8AJSQALwAfNx83e1QkIMVVfq1Ro2mFlZtXVEAZlaQOZAA-EuQTXwA3MmpIQCMAJzllcTV1M0A2gC6fZAZWVZ5iIXFZTSEXPLM4qwANJA6g0TMEZt1gjdOjtwMBgClfJBCB1TPkALYAIysNxuuwRSPaB2OpDRmOYkKAA)

So, It seems like the `BetterReturnType<>` here is the best version?
The only downside is that one failing test.
